ggml-cuda.cu

```diff
@@ -5059,6 +5240,173 @@ template <bool need_check> static __global__ void
+                const int j_bs = (col_y_0/8 + j/8);
+                const int k_bs = k0/4 + threadIdx.x;
+
+                const int j_tile = j + threadIdx.x/4;
+                const int k_tile = WARP_SIZE + threadIdx.x % 4;
+
+                cuda::memcpy_async(&tile_y_qs[j_tile*(WARP_SIZE + MMI8_PADDING_Y) + k_tile],
+                                &y_bs[j_bs*(nrows_y/16) + k_bs], 4, barriers[0]);
+            }
+        }
+
+#pragma unroll
+        for (int i0 = 0; i0 < mmi8_y; i0 += nwarps*MMI8_COPY_SIZE) {
+            const int x = MMI8_COPY_SIZE*(threadIdx.x % (WARP_SIZE/MMI8_COPY_SIZE));
+            const int y = MMI8_COPY_SIZE*threadIdx.y + threadIdx.x/(WARP_SIZE/MMI8_COPY_SIZE);
+
+            int row_x_eff = row_x_0 + i0 + y;
+            if (need_check_x) {
+                row_x_eff = min(row_x_eff, nrows_x-1); // to prevent out-of-bounds memory accesses
+            }
+
+            const int index_tile = (i0 + y)   * (WARP_SIZE + MMI8_PADDING_X)      + x;
+            const int index_qs   =  row_x_eff * (ncols_x/sizeof(int))        + k0 + x;
+            cuda::memcpy_async(&tile_x_qs[index_tile], &x_qs_low[index_qs], as, barriers[0]);
+        }
+
+#pragma unroll
+        for (int j0 = 0; j0 < mmi8_x; j0 += nwarps*MMI8_COPY_SIZE) {
+            const int x = MMI8_COPY_SIZE*(threadIdx.x % (WARP_SIZE/MMI8_COPY_SIZE));
+            const int y = MMI8_COPY_SIZE*threadIdx.y + threadIdx.x/(WARP_SIZE/MMI8_COPY_SIZE);
+
+            const int j = min(j0 + y, mmi8_x-1);
+
+            int col_y_eff = col_y_0 + j;
+            if (need_check_y) {
+                col_y_eff = min(col_y_eff, ncols_y-1); // to prevent out-of-bounds memory accesses
+            }
+
+            const int index_tile = j         * (WARP_SIZE + MMI8_PADDING_Y)      + x;
+            const int index_qs   = col_y_eff * (nrows_y/sizeof(int))        + k0 + x;
+            cuda::memcpy_async(&tile_y_qs[index_tile], &y_qs_low[index_qs], as, barriers[0]);
+        }
+
+        barriers[0].arrive_and_wait();
+
+#pragma unroll
+        for (int k = 0; k < 32; k += 16/sizeof(int)) {
+            const int ibs = ((k0/WARP_SIZE) % 4)*8 + k/4;
+
+            nvcuda::wmma::load_matrix_sync(
+                fa, (int8_t *) &tile_x_qs[threadIdx.y*(32*(WARP_SIZE + MMI8_PADDING_X)) + k],
+                (WARP_SIZE + MMI8_PADDING_X)*sizeof(int));
+
+#pragma unroll
+            for (int j = 0; j < mmi8_x; j += 8) {
+                const int bs = tile_y_qs[(j + ibs/4)*(WARP_SIZE + MMI8_PADDING_Y) + WARP_SIZE + ibs % 4];
+                nvcuda::wmma::load_matrix_sync(
+                    fb, (int8_t *) &tile_y_qs[j*(WARP_SIZE + MMI8_PADDING_Y) + k],
+                    (WARP_SIZE + MMI8_PADDING_Y)*sizeof(int));
+
+                nvcuda::wmma::fill_fragment(fc_tmp, 0);
+                nvcuda::wmma::mma_sync(fc_tmp, fa, fb, fc_tmp);
+#pragma unroll
+                for (int l = 0; l < 32*8/WARP_SIZE; ++l) {
+                    fc[j/8].x[l] += bs * fc_tmp.x[l];
```

Comment:
@jammm here I'm directly accessing the data that each thread holds and scaling it. Is this something that you can do on AMD?